Add the basic EDE (RFC8914) cases

TCY16 opened the pull request:
this branch is a split from #504 which implements all the easier EDE cases.

wtoorop committed:
Add EDE EDNS0 Option code

wtoorop committed:
Simple EDE example

Commit:
add locations for EDE errors

Commit:
Fix: use EDE codes with EDNS_OPT_APPEND_EDE

+ complete EDE codes

Commit:
Return EDE_DNSSEC_BOGUS when returning bogus answers

Commit:
Show reason when returning EDE_DNSSEC_BOGUS

Commit:
add variable bogus reason

Commit:
add local anwser blocked

Commit:
Revert "add local anwser blocked"

This reverts commit b3f60db18405f4395a5402e63a930f9a897a042b.

Commit:
Fix dont echo edns0 option list ...

when refusing to answer authoritatively.
Also remove TODO comments that were already done

Commit:
Merge branch 'master' into features/rfc8914-ede

Commit:
set up for tpkg test

Commit:
add localzones test

Commit:
add DNSSEC indeterminate EDE and DNAME expansion test

Commit:
add mesh bogus test, possible locations for more EDE and remove superfluous todo comments

Commit:
First step towards specific EDE DNSSEC errors

Commit:
add possible EDE spots

Commit:
Answer LDNS_EDE_RRSIGS_MISSING for normal answers with missing signatures

Commit:
add routine to do EDE on ACL blocked messages

Commit:
add forgotten compile error fixes from previous commit

Commit:
add routine to add EDE to ACL:refused at correct location

Commit:
change strncpy to memmove at @wcawijngaards' suggestion

Commit:
process @wcawijngaards' comments

Commit:
process @wcawijngaards' comments v2

Commit:
fix CH class response

Commit:
add QDCOUNT=0 to CHAOS query in ACL

Commit:
add EDE response to autotrust_init_fail test

Commit:
add EDE response to autotrust_init_failsig test

Commit:
add EDE responses to unittests

Commit:
add DNSKEY EDE code

Commit:
add more tests

Commit:
add test setup for DNSSEC EDEs

gthess committed:
- Introduce 'ede=<info-code>' and 'all_noedns' as options in the MATCH line for
  replay test packets.

Commit:
change unittests to match just the ede code

gthess committed:
- Fix testcode, 0 is a valid EDE INFO-CODE.

Commit:
add dnssec ede tests

Commit:
Merge branch 'master' into features/rfc8914-ede

Commit:
Fix merge bugs

Commit:
remove superfluous EDE left over from merge

Commit:
finish up adding validator EDEs and other TODOs and fix tests with more specific EDE codes

Commit:
add config options to test conf, fix local-zone EDE printing logic, and fix typo in the unbound.conf

Commit:
expand ede.tdir to do validator test for DNSKEY, RRSIG and NSEC missing

Commit:
add todo for tests and fix EDE codes for DNSKEY missing

Commit:
fix DNSSEC nsec-failure test

Commit:
fix rpl tests

Commit:
merge all basic EDE changes, merge with current master, and  resolve merge conflicts

Commit:
Merge branch 'master' into features/ede-basic

Commit:
remove unused function 'msgparse_check_edns_in_packet' left over from parent-split-off

Commit:
remove parent-split artifacts in localzone and fix tests correspondingly

Commit:
modify tdir test to ignore localzone tests and add cache snoop case

wtoorop commented on doc/unbound.conf.5.in:
```diff
@@ -1562,6 +1562,7 @@ Use this localzone type, regardless the type configured for the local-zone
 (both tagged and untagged) and regardless the type configured using
 access\-control\-tag\-action.
 .TP 5
+logic, and fix typo in the unbound.conf)
```
```suggestion
```

Some copy/paste mistake I think

Commit:
add EDE stale and modify rpl tests accordingly

Commit:
add forgotten rpl test for EDE stale

wtoorop commented on daemon/worker.c:
```diff
@@ -1029,15 +1045,124 @@ deny_refuse(struct comm_point* c, enum acl_access acl,
-		sldns_buffer_set_limit(c->buffer, LDNS_HEADER_SIZE);
-		sldns_buffer_write_at(c->buffer, 4, 
-			(uint8_t*)"\0\0\0\0\0\0\0\0", 8);
+		log_assert(LDNS_QDCOUNT(sldns_buffer_begin(c->buffer)) == 1);
```
Not a bug, but a suggestion.

```suggestion
		/* worker_check_request() above guarantees that the buffer contains at least a header and that qdcount ==  1 */
		log_assert(sldns_buffer_limit(pkt) >= LDNS_HEADER_SIZE && LDNS_QDCOUNT(sldns_buffer_begin(c->buffer)) == 1);
```

wtoorop commented on daemon/worker.c:
```diff
@@ -1029,15 +1045,124 @@ deny_refuse(struct comm_point* c, enum acl_access acl,
+		sldns_buffer_skip(c->buffer, LDNS_HEADER_SIZE); /* skip header */
+
+		if (!query_dname_len(c->buffer)) {
+			LDNS_QR_SET(sldns_buffer_begin(c->buffer));
```
Since the qname had a parse error, we should not return it so we return at least a **valid DNS error**.

```suggestion
			LDNS_QDCOUNT_SET(sldns_buffer_begin(c->buffer), 0);
			LDNS_ANCOUNT_SET(sldns_buffer_begin(c->buffer), 0);
			LDNS_NSCOUNT_SET(sldns_buffer_begin(c->buffer), 0);
			LDNS_ARCOUNT_SET(sldns_buffer_begin(c->buffer), 0);
			LDNS_QR_SET(sldns_buffer_begin(c->buffer));
```

wtoorop commented on daemon/worker.c:
```diff
@@ -1029,15 +1045,124 @@ deny_refuse(struct comm_point* c, enum acl_access acl,
+			sldns_buffer_flip(c->buffer);
+			return 1;
+		}
+		if (sldns_buffer_remaining(c->buffer) < 2 * sizeof(uint16_t)) {
```
```suggestion
		/* space available for query type and class? */
		if (sldns_buffer_remaining(c->buffer) < 2 * sizeof(uint16_t)) {
```

wtoorop commented on daemon/worker.c:
```diff
@@ -1029,15 +1045,124 @@ deny_refuse(struct comm_point* c, enum acl_access acl,
+                        LDNS_QR_SET(sldns_buffer_begin(c->buffer));
+                        LDNS_RCODE_SET(sldns_buffer_begin(c->buffer),
+				 LDNS_RCODE_FORMERR);
+                        sldns_buffer_flip(c->buffer);
```
It's an invalid query, we should not return an invalid error.
```suggestion
			LDNS_QDCOUNT_SET(sldns_buffer_begin(c->buffer), 0);
			LDNS_ANCOUNT_SET(sldns_buffer_begin(c->buffer), 0);
			LDNS_NSCOUNT_SET(sldns_buffer_begin(c->buffer), 0);
			LDNS_ARCOUNT_SET(sldns_buffer_begin(c->buffer), 0);
			sldns_buffer_set_position(c->buffer, LDNS_HEADER_SIZE);
                        sldns_buffer_flip(c->buffer);
```

I also notice mixed usage of tabs and 8 spaces around this area. FYI I've used tabs in the suggestion.

wtoorop commented on daemon/worker.c:
```diff
@@ -1029,15 +1045,124 @@ deny_refuse(struct comm_point* c, enum acl_access acl,
+			LDNS_ARCOUNT_SET(sldns_buffer_begin(c->buffer), 0);
+			sldns_buffer_flip(c->buffer);
+			return 1;
+		}
```
Why check for IN? Anyway, if we would REFUSE that immediately, your returned error would have trailing data too.

```suggestion
sldns_buffer_skip(c->buffer, (ssize_t)sizeof(uint16_t)); /* skip qclass */
```

Note that below you would not have trailing data because you are not resetting the QDCOUNT.

wtoorop commented on daemon/worker.c:
```diff
@@ -1029,15 +1045,124 @@ deny_refuse(struct comm_point* c, enum acl_access acl,
+				sldns_buffer_flip(c->buffer);
+				return 1;
+			}
+		}
```
Then at least get rid of the broken answer and NS section to at least return a valid error:

```suggestion
		/* Skip through the RR records */
		if(LDNS_ANCOUNT(sldns_buffer_begin(c->buffer)) != 0 ||
			LDNS_NSCOUNT(sldns_buffer_begin(c->buffer)) != 0) {
			if(!skip_pkt_rrs(c->buffer, 
				((int)LDNS_ANCOUNT(sldns_buffer_begin(c->buffer)))+
				((int)LDNS_NSCOUNT(sldns_buffer_begin(c->buffer))))) {
				LDNS_RCODE_SET(sldns_buffer_begin(c->buffer),
					LDNS_RCODE_FORMERR);
				LDNS_ANCOUNT_SET(sldns_buffer_begin(c->buffer), 0);
				LDNS_NSCOUNT_SET(sldns_buffer_begin(c->buffer), 0);
				LDNS_ARCOUNT_SET(sldns_buffer_begin(c->buffer), 0);
				sldns_buffer_set_position(c->buffer, opt_rr_mark);
				sldns_buffer_flip(c->buffer);
				return 1;
			}
		}
```

wtoorop commented on daemon/worker.c:
```diff
@@ -1029,15 +1045,124 @@ deny_refuse(struct comm_point* c, enum acl_access acl,
+			LDNS_RCODE_SET(sldns_buffer_begin(c->buffer),
+				LDNS_RCODE_FORMERR);
+			return 1;
+		}
```
Return REFUSED instead of FORMERR because the packet may well be just valid, and a suggestion to not spend cycles on dname's we don't care about. Also you forgot to clear the counters, set the position to right after the query and flip the buffer.

```suggestion
		/* Do we have a valid OPT RR here? If not return REFUSED (could be a valid TSIG or something so no FORMERR) */
		/* domain name must be the root of length 1. */
		if(sldns_buffer_remaining(c->buffer) < 1 || *sldns_buffer_current(c->buffer) != 0) {
			LDNS_ANCOUNT_SET(sldns_buffer_begin(c->buffer), 0);
			LDNS_NSCOUNT_SET(sldns_buffer_begin(c->buffer), 0);
			LDNS_ARCOUNT_SET(sldns_buffer_begin(c->buffer), 0);
			sldns_buffer_set_position(c->buffer, opt_rr_mark);
			sldns_buffer_flip(c->buffer);
			return 1;
		} else
			sldns_buffer_skip(c->buffer, 1); /* skip root label */
```

wtoorop commented on daemon/worker.c:
```diff
@@ -1029,15 +1045,124 @@ deny_refuse(struct comm_point* c, enum acl_access acl,
+				LDNS_RCODE_FORMERR);
+			sldns_buffer_flip(c->buffer);
+			return 1;
+		}
```
I don't see why this would be a FORMERR, changing to REFUSED instead. Also you forgot to clear counters and reset the position.

```suggestion
		if(sldns_buffer_remaining(c->buffer) < 2 ||
			sldns_buffer_read_u16(c->buffer) != LDNS_RR_TYPE_OPT) {
			LDNS_ANCOUNT_SET(sldns_buffer_begin(c->buffer), 0);
			LDNS_NSCOUNT_SET(sldns_buffer_begin(c->buffer), 0);
			LDNS_ARCOUNT_SET(sldns_buffer_begin(c->buffer), 0);
			sldns_buffer_flip(c->buffer);
			return 1;
		}
```

Commit:
add EDE stale for mesh, change rpl tests accordingly

wtoorop reviewed (changes requested):
Review of just the ACL refused stuff.
The rest is coming in separate review batches.

wtoorop commented on util/config_file.c:
```diff
@@ -373,6 +373,7 @@ config_create(void)
 	cfg->ipset_name_v4 = NULL;
 	cfg->ipset_name_v6 = NULL;
 #endif
+	cfg->local_data_do_ede = 0;
```
Part of the not so simple EDE cases?

```suggestion
```

TCY16 replied:
resolved in 71e7bbf

wtoorop commented on util/config_file.c:
```diff
@@ -790,7 +791,8 @@ int config_set_option(struct config_file* cfg, const char* opt,
 		cfg->out_ifs = oi;
-	} else {
+	} else S_YNO("local_data_do_ede:", local_data_do_ede)
+	else {
```
Part of the not so simple EDE cases?

```suggestion
	} else  {
```

TCY16 replied:
resolved in 71e7bbf

wtoorop commented on util/config_file.h:
```diff
@@ -668,6 +669,9 @@ struct config_file {
 #endif
+
+	/** should local_data result in EDE (RFC8914) code inclusion? */
+	int local_data_do_ede;
```
Part of the not so simple EDE cases?

```suggestion
```

TCY16 replied:
resolved in 71e7bbf

wtoorop commented on services/localzone.c:
```diff
@@ -1324,11 +1324,55 @@ local_encode(struct query_info* qinfo, struct module_env* env,
+	return 1;
+}
+
+
```
Part of the not so simple EDE cases?

```suggestion
```

wtoorop commented on services/localzone.c:
```diff
@@ -1580,8 +1627,8 @@ local_data_answer(struct local_zone* z, struct module_env* env,
-	return local_encode(qinfo, env, edns, repinfo, buf, temp, lr->rrset, 1,
-		LDNS_RCODE_NOERROR);
+	return local_encode_ede(qinfo, env, edns, repinfo, buf, temp, lr->rrset, 1,
+		LDNS_RCODE_NOERROR, LDNS_EDE_FORGED_ANSWER, "");
```
Part of the not so simple EDE cases?

```suggestion
	return local_encode(qinfo, env, edns, repinfo, buf, temp, lr->rrset, 1,
		LDNS_RCODE_NOERROR, -1, NULL);
```

wtoorop commented on services/localzone.c:
```diff
@@ -1702,12 +1750,13 @@ local_zones_zone_answer(struct local_zone* z, struct module_env* env,
-			return local_encode(qinfo, env, edns, repinfo, buf, temp,
-				&lrr, 1, LDNS_RCODE_NOERROR);
+			return local_encode_ede(qinfo, env, edns, repinfo, buf, temp,
+				&lrr, 1, LDNS_RCODE_NOERROR, LDNS_EDE_FORGED_ANSWER, "");
```
Part of the not so simple EDE cases?

```suggestion
			return local_encode(qinfo, env, edns, repinfo, buf, temp,
				&lrr, 1, LDNS_RCODE_NOERROR, -1, NULL);
```

wtoorop commented on services/localzone.c:
```diff
@@ -1324,11 +1324,55 @@ local_encode(struct query_info* qinfo, struct module_env* env,
 	struct edns_data* edns, struct comm_reply* repinfo, sldns_buffer* buf,
-	struct regional* temp, int rcode, int r)
+	struct regional* temp, int rcode, int r, sldns_ede_code ede_code,
+	const char* ede_txt)
```
Variables `ede_code` and `ede_txt` are not used in the function body!

TCY16 replied:
resolved in 71e7bbf

wtoorop commented on services/localzone.c:
```diff
@@ -1638,7 +1685,8 @@ local_zones_zone_answer(struct local_zone* z, struct module_env* env,
 		local_error_encode(qinfo, env, edns, repinfo, buf, temp,
-			LDNS_RCODE_REFUSED, (LDNS_RCODE_REFUSED|BIT_AA));
+			LDNS_RCODE_REFUSED, (LDNS_RCODE_REFUSED|BIT_AA),
+			LDNS_EDE_BLOCKED, "");
```
Part of the not so simple EDE cases?

```suggestion
			LDNS_RCODE_REFUSED, (LDNS_RCODE_REFUSED|BIT_AA), -1, NULL);
```

wtoorop commented on services/localzone.c:
```diff
@@ -1663,8 +1711,8 @@ local_zones_zone_answer(struct local_zone* z, struct module_env* env,
-		local_error_encode(qinfo, env, edns, repinfo, buf, temp, rcode,
-			(rcode|BIT_AA));
+		local_error_encode(qinfo, env, edns, repinfo, buf, temp,
+			rcode, (rcode|BIT_AA), LDNS_EDE_BLOCKED, "");
```
Part of the not so simple EDE cases?
```suggestion
		local_error_encode(qinfo, env, edns, repinfo, buf, temp,
			rcode, (rcode|BIT_AA), -1, NULL);
```

wtoorop reviewed (changes requested):
Please let  `local_error_encode()` use the `ede` and `ede_txt` parameters or create another function for it (`local_error_encode_ede()`?).

wtoorop commented on daemon/worker.c:
```diff
@@ -484,6 +484,8 @@ answer_norec_from_cache(struct worker* worker, struct query_info* qinfo,
 				worker->env.now_tv))
 					return 0;
+			EDNS_OPT_LIST_APPEND_EDE(&edns->opt_list_out,
+				worker->scratchpad, LDNS_EDE_DNSSEC_BOGUS, "");
```
Okay for now, but it would eventually be nice if a reason could somehow be extracted or stored in the cache.
Can you make a note for this?

TCY16 replied:
Noted

wtoorop commented on daemon/worker.c:
```diff
@@ -499,6 +501,9 @@ answer_norec_from_cache(struct worker* worker, struct query_info* qinfo,
 		case sec_status_indeterminate:
+			EDNS_OPT_LIST_APPEND_EDE(&edns->opt_list_out,
+				worker->scratchpad, LDNS_EDE_DNSSEC_INDETERMINATE, "");
+			/* fallthrough */
```
This will add and EDE to all answers if no trust anchor is configured. Not a good idea! We definitely want no EDE's added to positive answers when not explicitly configured like that.
This should be turned on explicitly by a config option, and/or the queries could perhaps ask for it by sending an EDE option in the query? That last idea should be discussed in IETF b.t.w. Can you make a note for that? Thanks!

```suggestion
			EDNS_OPT_LIST_APPEND_EDE(&edns->opt_list_out,
				worker->scratchpad, LDNS_EDE_DNSSEC_INDETERMINATE, "");
			/* fallthrough */
```

TCY16 replied:
removed and noted

wtoorop commented on daemon/worker.c:
```diff
@@ -654,6 +659,8 @@ answer_from_cache(struct worker* worker, struct query_info* qinfo,
 			worker->env.now_tv))
 			goto bail_out;
+		EDNS_OPT_LIST_APPEND_EDE(&edns->opt_list_out,
+			worker->scratchpad, LDNS_EDE_DNSSEC_BOGUS, "");
```
Okay for now, but it would eventually be nice if a reason could somehow be extracted or stored in the cache.
Can you make a note for this?

TCY16 replied:
Noted

wtoorop commented on daemon/worker.c:
```diff
@@ -716,15 +723,22 @@ answer_from_cache(struct worker* worker, struct query_info* qinfo,
-			qinfo, id, flags, edns);
+	} else {
+		if (*is_expired_answer == 1) {
+			EDNS_OPT_LIST_APPEND_EDE(&edns->opt_list_out,
+				worker->scratchpad, LDNS_EDE_STALE_ANSWER, "");
```
This needs to be turned on explicitly in the config file!

TCY16 replied:
Added as config option

wtoorop commented on daemon/worker.c:
```diff
@@ -1500,6 +1630,9 @@ worker_handle_request(struct comm_point* c, void* arg, int error,
 		}
+
+		// @TODO Extended DNS Error Code 13 - Cached Error? place not clear
+
```
Well... not here, so please leave the note TODO elsewhere
```suggestion
```

TCY16 replied:
removed

wtoorop commented on services/mesh.c:
```diff
@@ -1290,6 +1290,22 @@ mesh_send_reply(struct mesh_state* m, int rcode, struct reply_info* rep,
 		}
+		/* Send along EDE BOGUS EDNS0 option when answer is bogus */
+		if(rcode == LDNS_RCODE_SERVFAIL &&
+			m->s.env->need_to_validate && (!(r->qflags&BIT_CD) ||
+			m->s.env->cfg->ignore_cd) && rep &&
```
Should an EDE code be included when checking is disabled?
I tried with quad1 which is the only public resolver supporting EDE I know of, and they ignore cd bit!
Maybe we should seek advice at DNSOP?

TCY16 replied:
> Maybe we should seek advice at DNSOP?

I like this idea!

gthess replied:
@rijswijk suggested to have this functionality configured. Default would be to not attach if CD is set.

Edit: Through discussion with @TCY16 and @gthess we've decided that we will leave this for when we implement storing EDE in the cached records.

wtoorop commented on services/mesh.c:
```diff
@@ -2042,6 +2060,12 @@ mesh_serve_expired_callback(void* arg)
+			edns_opt_list_append_ede(&r->edns.opt_list_out,
+				mstate->s.region, LDNS_EDE_STALE_ANSWER, "");
+		}
+
```
Again, this should be configured explicitly in the config file.
Another idea, it would be cool if we could mention how stale the reply is in the EDE text. Can you make a note? Thanks!

```
serve-expired-ede: <yes or no>

  Extended DNS Error Code 3 - Stale Answer as EDNS0 option to  the response provided that do-ede is yes.
  default is no.
```

TCY16 replied:
Added as config option

wtoorop reviewed (changes requested):
Mesh and worker changes

wtoorop commented on validator/validator.c:
```diff
@@ -2676,7 +2697,7 @@ process_ds_response(struct module_qstate* qstate, struct val_qstate* vq,
  */
 static void
-process_dnskey_response(struct module_qstate* qstate, struct val_qstate* vq,
+  process_dnskey_response(struct module_qstate* qstate, struct val_qstate* vq,
```
2 spaces?

```suggestion
process_dnskey_response(struct module_qstate* qstate, struct val_qstate* vq,
```

wtoorop commented on validator/validator.c:
```diff
@@ -2693,6 +2715,9 @@ process_dnskey_response(struct module_qstate* qstate, struct val_qstate* vq,
 			"DNSKEY query.");
+
+		reason_bogus = LDNS_EDE_DNSKEY_MISSING;
+
```
No need if you do this with `errinf_ede()` later on, right?

```suggestion
```

wtoorop commented on validator/validator.c:
```diff
@@ -2707,7 +2732,7 @@ process_dnskey_response(struct module_qstate* qstate, struct val_qstate* vq,
 			/* key_entry is NULL for failure in Validate */
 		}
-		errinf(qstate, "No DNSKEY record");
+		errinf_ede(qstate, "No DNSKEY record", reason_bogus);
```
```suggestion
		errinf_ede(qstate, "No DNSKEY record", LDNS_EDE_DNSKEY_MISSING);
```

wtoorop reviewed (commented):
Review of validator.c

TCY16 committed:
Apply suggestions to ACL code

Co-authored-by: Willem Toorop <willem@nlnetlabs.nl>

TCY16 committed:
Cleanup of the not so simple EDE cases

Co-authored-by: Willem Toorop <willem@nlnetlabs.nl>

Commit:
remove more not-so-simple EDE cases in local_zone and the coresponding tests, and config_file.*, fix ACL by difusing CH queries

Commit:
add serve-expired-ede as a config option and processed review comments for mesh and worker

Commit:
process review comments on validator

Commit:
add verbose EDE attach logging, change simple tpk tests to .rpls, and change EDE type definition in local_zone:local_error_encode

TCY16 committed:
fix typo in doc/unbound.conf.5.in

Co-authored-by: Willem Toorop <willem@nlnetlabs.nl>

Commit:
remove tpkg test cases that became rpls

Commit:
Merge branch 'master' into features/ede-basic

Commit:
add EDE code in EDNS list, stop-gap nsid_bogus rpl, and fix error in validator where reason_bogus was not initialized

Commit:
add global do-ede config option which is checked when attaching an EDE. Change attachment points, rpls, and manpage accordingly

wtoorop commented on daemon/worker.c:
```diff
@@ -1029,32 +1053,165 @@ deny_refuse(struct comm_point* c, enum acl_access acl,
+			LDNS_QR_SET(sldns_buffer_begin(c->buffer));
+			LDNS_RCODE_SET(sldns_buffer_begin(c->buffer),
+				LDNS_RCODE_FORMERR);
+			sldns_buffer_flip(c->buffer);
```
`query_dname_len()` can leave the buffer position on an unknown position somewhere in the query name (on error) so you should reset to just after the header.

```suggestion
			sldns_buffer_set_position(c->buffer, LDNS_HEADER_SIZE);
			sldns_buffer_flip(c->buffer);
```

wtoorop commented on services/mesh.c:
```diff
@@ -1290,6 +1290,24 @@ mesh_send_reply(struct mesh_state* m, int rcode, struct reply_info* rep,
+			if (m->s.env->cfg->do_ede) {
+				edns_opt_list_append_ede(&r->edns.opt_list_out, m->s.region,
+					reason_bogus, reason);
+			}
```
The do-ede test can go one level up, so you don't have to do the work (allocation of reason_bogus in `errinf_to_readon_bogus()`) when it is discarded anyway.
```suggestion
		if(m->s.env->cfg->do_ede && rcode == LDNS_RCODE_SERVFAIL &&
			m->s.env->need_to_validate && (!(r->qflags&BIT_CD) ||
			m->s.env->cfg->ignore_cd) && rep &&
			(rep->security <= sec_status_bogus ||
			rep->security == sec_status_secure_sentinel_fail)) {

			char *reason = m->s.env->cfg->val_log_level >= 2
			             ? errinf_to_str_bogus(&m->s) : NULL;
			sldns_ede_code reason_bogus = rep->reason_bogus != LDNS_EDE_DNSSEC_BOGUS
			                            ? rep->reason_bogus : errinf_to_reason_bogus(&m->s);

			edns_opt_list_append_ede(&r->edns.opt_list_out, m->s.region,
					reason_bogus, reason);
```

wtoorop commented on validator/val_utils.c:
```diff
@@ -406,16 +416,28 @@ val_verify_rrset_entry(struct module_env* env, struct val_env* ve,
+	return verify_dnskeys_with_ds_rr_ede( env, ve, dnskey_rrset, ds_rrset,
+		ds_idx, reason, NULL, qstate);
+}
+#endif
```
Some sort of development leftover? Just remove I suppose...

```suggestion
```

wtoorop commented on validator/val_sigcrypt.c:
```diff
@@ -1386,6 +1436,14 @@ check_dates(struct val_env* ve, uint32_t unow,
+			 * attempted to perform DNSSEC validation, but no
+			 * signatures are presently valid and at least some are
+			 * not yet valid. */
+			*reason_bogus = LDNS_EDE_SIGNATURE_NOT_YET_VALID;
```
We need a new Extended DNS Error Code for the specific case that the expiry time of a signature is before the inception time. We request the following entry to be added to the ["Extended DNS Error Codes" on the "Domain Name System (DNS) Parameters" web page](https://www.iana.org/assignments/dns-parameters/dns-parameters.xhtml#extended-dns-error-codes):

| INFO-CODE      | Purpose                        | Reference        |
| -------------- | ------------------------------ | ---------------- |
| 25        | Signature Expired before Valid | [Comment in Pull Request for Extended DNS Errors in Unbound](https://github.com/NLnetLabs/unbound/pull/604#discussion_r802678343) |

This  Extended DNS Error code can be returned when the resolver attempted to perform DNSSEC validation, but no signatures are presently valid and at least some have an expiry date before the inception date.

wtoorop commented on validator/val_sigcrypt.c:
```diff
@@ -1492,19 +1568,25 @@ dnskey_verify_rrset_sig(struct regional* region, sldns_buffer* buf,
 		verbose(VERB_QUERY, "verify: dnskey has wrong key protocol");
 		*reason = "dnskey has wrong protocolnumber";
+		if(reason_bogus)
+			*reason_bogus = LDNS_EDE_DNSSEC_BOGUS;
```
We need a new Extended DNS Error code for the specific case where the DNSKEY protocol number is not 3 (as specified in [RFC4034](https://datatracker.ietf.org/doc/html/rfc4034#section-2.1.2)). We request the following entry to be added to the ["Extended DNS Error Codes" on the "Domain Name System (DNS) Parameters" web page](https://www.iana.org/assignments/dns-parameters/dns-parameters.xhtml#extended-dns-error-codes):

| INFO-CODE      | Purpose                        | Reference        |
| -------------- | ------------------------------ | ---------------- |
| 31        | DNSSEC Unknown Protocol| [Comment in Pull Request for Extended DNS Errors in Unbound](https://github.com/NLnetLabs/unbound/pull/604#discussion_r802735845) |

This Extended DNS Error code can be returned when the resolver attempted to perform DNSSEC validation, but the wrong value in the DNSKEY Protocol Field was found.

wtoorop commented on validator/val_sigcrypt.c:
```diff
@@ -1492,19 +1568,25 @@ dnskey_verify_rrset_sig(struct regional* region, sldns_buffer* buf,
 		verbose(VERB_QUERY, "verify: signature too short");
 		*reason = "signature too short";
+		if(reason_bogus)
+			*reason_bogus = LDNS_EDE_DNSSEC_BOGUS;
```
We need a new Extended DNS Error code for the specific case where an attached signature is either smaller or larger than expected for the specified algorithm. We request the following entry to be added to the ["Extended DNS Error Codes" on the "Domain Name System (DNS) Parameters" web page](https://www.iana.org/assignments/dns-parameters/dns-parameters.xhtml#extended-dns-error-codes):

| INFO-CODE      | Purpose                        | Reference        |
| -------------- | ------------------------------ | ---------------- |
| 26        | Signature Wrong Size| [Comment in Pull Request for Extended DNS Errors in Unbound](https://github.com/NLnetLabs/unbound/pull/604#discussion_r802737095) |

This Extended DNS Error code can be returned when the resolver attempted to perform DNSSEC validation, but no undeformed signature was found.

wtoorop commented on validator/val_sigcrypt.c:
```diff
@@ -1514,17 +1596,23 @@ dnskey_verify_rrset_sig(struct regional* region, sldns_buffer* buf,
 		verbose(VERB_QUERY, "verify: malformed signer name");
 		*reason = "signer name malformed";
+		if(reason_bogus)
+			*reason_bogus = LDNS_EDE_DNSSEC_BOGUS;
```
We need a new Extended DNS Error code for the specific case where the signer name is malformed. We request the following entry to be added to the ["Extended DNS Error Codes" on the "Domain Name System (DNS) Parameters" web page](https://www.iana.org/assignments/dns-parameters/dns-parameters.xhtml#extended-dns-error-codes):

| INFO-CODE      | Purpose                        | Reference        |
| -------------- | ------------------------------ | ---------------- |
| 27        | Malformed Signer Name| [Comment in Pull Request for Extended DNS Errors in Unbound](https://github.com/NLnetLabs/unbound/pull/604#discussion_r802737749) |

This Extended DNS Error code can be returned when the resolver attempted to perform DNSSEC validation, but  the Signer's Name Field in the signature contains a malformed signer name, which can be either too short or too long for example.

wtoorop commented on validator/val_sigcrypt.c:
```diff
@@ -1514,17 +1596,23 @@ dnskey_verify_rrset_sig(struct regional* region, sldns_buffer* buf,
 		verbose(VERB_QUERY, "verify: signer name is off-tree");
 		*reason = "signer name off-tree";
+		if(reason_bogus)
+			*reason_bogus = LDNS_EDE_DNSSEC_BOGUS;
```
We need a new Extended DNS Error code for the specific case where the signer name is outside of the zone that the signature is for. We request the following entry to be added to the ["Extended DNS Error Codes" on the "Domain Name System (DNS) Parameters" web page](https://www.iana.org/assignments/dns-parameters/dns-parameters.xhtml#extended-dns-error-codes):

| INFO-CODE      | Purpose                        | Reference        |
| -------------- | ------------------------------ | ---------------- |
| 28        | Signer Name Out Of Zone| [Comment in Pull Request for Extended DNS Errors in Unbound](https://github.com/NLnetLabs/unbound/pull/604#discussion_r802738414) |

This Extended DNS Error code can be returned when the resolver attempted to perform DNSSEC validation, but  the Signer's Name Field in the signature does not contain the zone name of the covered RRset.

wtoorop commented on validator/val_sigcrypt.c:
```diff
@@ -1514,17 +1596,23 @@ dnskey_verify_rrset_sig(struct regional* region, sldns_buffer* buf,
 		verbose(VERB_QUERY, "verify: too short, no signature data");
 		*reason = "signature too short, no signature data";
+		if(reason_bogus)
+			*reason_bogus = LDNS_EDE_DNSSEC_BOGUS;
```
Potential new EDE code: signature wrong size

TCY16 replied:
Settled in the [comment above](https://github.com/NLnetLabs/unbound/pull/604#discussion_r802737095).

wtoorop commented on validator/val_sigcrypt.c:
```diff
@@ -1537,6 +1625,8 @@ dnskey_verify_rrset_sig(struct regional* region, sldns_buffer* buf,
 			dnskey->rk.dname, 0, 0);
 		*reason = "signer name mismatches key name";
+		if(reason_bogus)
+			*reason_bogus = LDNS_EDE_DNSSEC_BOGUS;
```
Note to myself: How can this happen? Is the key only selected by id?

wtoorop commented on validator/val_sigcrypt.c:
```diff
@@ -1545,25 +1635,33 @@ dnskey_verify_rrset_sig(struct regional* region, sldns_buffer* buf,
 		verbose(VERB_QUERY, "verify: wrong type covered");
 		*reason = "signature covers wrong type";
+		if(reason_bogus)
+			*reason_bogus = LDNS_EDE_DNSSEC_BOGUS;
```
@wcawijngaards Can this actually happen? How?

wcawijngaards replied:
No, it would be an error in the parser, that put an RRSIG of a different covered type into the RRset. Thus it is defensive code.

wtoorop commented on validator/val_sigcrypt.c:
```diff
@@ -1545,25 +1635,33 @@ dnskey_verify_rrset_sig(struct regional* region, sldns_buffer* buf,
 		verbose(VERB_QUERY, "verify: labelcount out of range");
 		*reason = "signature labelcount out of range";
+		if(reason_bogus)
+			*reason_bogus = LDNS_EDE_DNSSEC_BOGUS;
```
We need a new Extended DNS Error code for the specific case where has the wrong label count for the signature owner name. We request the following entry to be added to the ["Extended DNS Error Codes" on the "Domain Name System (DNS) Parameters" web page](https://www.iana.org/assignments/dns-parameters/dns-parameters.xhtml#extended-dns-error-codes):

| INFO-CODE      | Purpose                        | Reference        |
| -------------- | ------------------------------ | ---------------- |
| 29        | Signature Wrong Labelcount| [Comment in Pull Request for Extended DNS Errors in Unbound](https://github.com/NLnetLabs/unbound/pull/604#discussion_r802743293) |

This Extended DNS Error code can be returned when the resolver attempted to perform DNSSEC validation, but  the number of labels in the Signature Labels Field is incorrect.

wtoorop commented on validator/val_sigcrypt.h:
```diff
@@ -285,6 +292,13 @@ enum sec_status dnskey_verify_rrset(struct module_env* env,
+	size_t dnskey_idx, char** reason, sldns_ede_code *reason_bogus,
+	sldns_pkt_section section, struct module_qstate* qstate);
+
+#if 0
```
@TCY16 Could you review all te left over commenting out with `#if 0` and remove those sections.
I cannot do it with a suggestion as it also included code not in the commit.

wtoorop commented on validator/val_utils.c:
```diff
@@ -560,7 +593,7 @@ val_verify_DNSKEY_with_DS(struct module_env* env, struct val_env* ve,
 		verbose(VERB_ALGO, "No usable DS records were found -- "
 			"treating as insecure.");
-		return sec_status_insecure;
+			return sec_status_insecure;
```
Accidentally added a tab?
```suggestion
		return sec_status_insecure;
```

wtoorop commented on validator/validator.c:
```diff
@@ -2397,7 +2418,8 @@ ds_response_to_ke(struct module_qstate* qstate, struct val_qstate* vq,
 		verbose(VERB_DETAIL, "DS response was error, thus bogus");
 		errinf(qstate, rc);
-		errinf(qstate, "no DS");
+		errinf_ede(qstate, "no DS", reason_bogus);
```
It doesn't make sense to use the variable `reason_bogus` here, because it cannot be anything else than `LDNS_EDE_DNSSEC_BOGUS` anyway. But was about `LDNS_EDE_NO_REACHABLE_AUTHORITY` or `LDNS_EDE_NETWORK_ERROR`?

```suggestion
		errinf_ede(qstate, "no DS", LDNS_EDE_NETWORK_ERROR);
```

wtoorop commented on validator/validator.c:
```diff
@@ -2411,17 +2433,17 @@ ds_response_to_ke(struct module_qstate* qstate, struct val_qstate* vq,
 			log_warn("internal error: POSITIVE DS response was "
 				"missing DS.");
-			errinf(qstate, "no DS record");
+			errinf_ede(qstate, "no DS record", reason_bogus);
```
Anything but `reason_bogus`

```suggestion
			errinf_ede(qstate, "no DS record", LDNS_EDE_DNSSEC_BOGUS);
```

wtoorop commented on validator/validator.c:
```diff
@@ -1635,14 +1645,15 @@ processInit(struct module_qstate* qstate, struct val_qstate* vq,
 		/* key is bad, chain is bad, reply is bogus */
 		errinf_dname(qstate, "key for validation", vq->key_entry->name);
-		errinf(qstate, "is marked as invalid");
+		errinf_ede(qstate, "is marked as invalid", LDNS_EDE_DNSSEC_BOGUS);
```
This will cause a peculiar text message in the EDE I think. Do we need a `errinf_dname_ede()`?

wtoorop commented on validator/validator.c:
```diff
@@ -1848,7 +1860,9 @@ processValidate(struct module_qstate* qstate, struct val_qstate* vq,
-		errinf(qstate, "while building chain of trust");
+		vq->chase_reply->reason_bogus = LDNS_EDE_DNSSEC_BOGUS;
+		errinf_ede(qstate, "while building chain of trust",
+			LDNS_EDE_DNSKEY_MISSING);
```
Why the discrepancy between setting `vq->chase_reply->reason_bogus` and `errinf_ede()`?

```suggestion
		vq->chase_reply->reason_bogus = LDNS_EDE_DNSKEY_MISSING;
		errinf_ede(qstate, "while building chain of trust",
			LDNS_EDE_DNSKEY_MISSING);
```

wtoorop commented on doc/unbound.conf.5.in:
```diff
@@ -1761,6 +1765,11 @@ option can be used multiple times. The most specific match will be used.
+.B do\-ede: \fI<yes or no>
+If enabled, Unbound will respond with Extended DNS Error codes (RFC8914).
+These EDEs attachinformative error messages to a response for various
+errors. Default is "no".
```
```suggestion
errors. Default is "no".

When the \fBval-log-level\fR option is also set to \fB2\fR, responses with Extended DNS Errors concerning DNSSEC failures that are not served from cache, will also contain a descriptive text message about the reason for the failure.
```

wtoorop reviewed (changes requested):
Thanks Tom! Very nice! I like the new ACL rpl's! Well done. There are a few remaining nits.

Also, I have a remaining concern, I saw there are quite a few assignments of a value of `sec_status_bogus` to the `security` attribute of a `struct replay_info` without also the `reason_bogus` attribute being set... I think these should be revisited and if unsure at least assign `LDNS_EDE_DNSSEC_BOGUS` to the `reason_bogus` attribute.

gthess committed:
- Introduce 'ede=any' as option in the MATCH line for replay test
  packets.
- Finish EDE matching code; when matching EDE snip the option out of the
  OPT record to facilitate other matching rules.

gthess committed:
- Update replay tests for new EDE matching logic.

Commit:
add debug statement to make test which gives the output of the failed test

Commit:
process @wtoorop's review comments v1

Commit:
add missing sec_status_bogus EDEs

TCY16 commented on validator/validator.c:
```diff
@@ -963,6 +965,7 @@ validate_nodata_response(struct module_env* env, struct val_env* ve,
 		if(verbosity >= VERB_ALGO)
 			log_dns_msg("Failed NODATA", qchase, chase_reply);
 		chase_reply->security = sec_status_bogus;
+		chase_reply->reason_bogus = LDNS_EDE_DNSSEC_BOGUS;
```
We need a new Extended DNS Error code for the specific case where the response gives insufficient proof. We request the following entry to be added to the ["Extended DNS Error Codes" on the "Domain Name System (DNS) Parameters" web page](https://www.iana.org/assignments/dns-parameters/dns-parameters.xhtml#extended-dns-error-codes):

| INFO-CODE      | Purpose                        | Reference        |
| -------------- | ------------------------------ | ---------------- |
| 30        | DNSSEC Insufficient NSEC Proof| [Comment in Pull Request for Extended DNS Errors in Unbound](https://github.com/NLnetLabs/unbound/pull/604#discussion_r803797225) |

This Extended DNS Error code can be returned when the resolver attempted to perform DNSSEC validation, but  the signed response does not have valid NSEC proof.

gthess committed:
- Show the replay test that failed again after the output.

gthess committed:
- Makefile commands should be oneliners.

gthess committed:
- Add missing '\'.

gthess committed:
- Fix for EDE test code: initialise variable.

gthess committed:
- Fix for EDE test code: Restore the original packet for each entry.

gthess committed:
- Fix typo in test.

gthess committed:
- Fix doc string after feedback from doxygen.

gthess committed:
Merge branch 'master' into features/ede-basic

gthess committed:
- Fix EDE test code: reset the original packet on return.

Commit:
change logic in validator.c:processValidate to not overwrite previously set reason_bogus

Commit:
remove superflous _ede from functions

Commit:
add missing do-ede: yes to tpkg conf

wcawijngaards commented on daemon/worker.c:
```diff
@@ -1030,32 +1054,166 @@ deny_refuse(struct comm_point* c, enum acl_access acl,
+		log_assert(sldns_buffer_limit(c->buffer) >= LDNS_HEADER_SIZE
+			&& LDNS_QDCOUNT(sldns_buffer_begin(c->buffer)) == 1);
+
+		sldns_buffer_clear(c->buffer); /* reset write limit */
```
```suggestion
```
The buffer clear has to not be here, because otherwise at line 1136 and 1146 it could start reading from previous packets still from earlier content in the buffer space. This happens because those if statements try to check read lengths, but instead get the write length because of the buffer clear. Leakage of previous packets is a bad type of information disclosure vuln. The fix is to remove the buffer clear from here and insert it at line 1156, there you switch from reading the incoming packet to writing the outgoing packet.

wcawijngaards commented on daemon/worker.c:
```diff
@@ -1030,32 +1054,166 @@ deny_refuse(struct comm_point* c, enum acl_access acl,
+		 */
+		LDNS_ANCOUNT_SET(sldns_buffer_begin(c->buffer), 0);
+		LDNS_NSCOUNT_SET(sldns_buffer_begin(c->buffer), 0);
+		sldns_buffer_set_position(c->buffer, opt_rr_mark);
```
```suggestion
		LDNS_ANCOUNT_SET(sldns_buffer_begin(c->buffer), 0);
		LDNS_NSCOUNT_SET(sldns_buffer_begin(c->buffer), 0);
		sldns_buffer_clear(c->buffer); /* reset write limit */
		sldns_buffer_set_position(c->buffer, opt_rr_mark);
```
Insert the buffer_clear call here, where it switches from read of the old packet into write of the new packet. It is now the new sizes want to be checked for, for inserting the new content length of the EDE data.

wcawijngaards commented on daemon/worker.c:
```diff
@@ -1030,32 +1054,166 @@ deny_refuse(struct comm_point* c, enum acl_access acl,
+			LDNS_ANCOUNT_SET(sldns_buffer_begin(c->buffer), 0);
+			LDNS_NSCOUNT_SET(sldns_buffer_begin(c->buffer), 0);
+			LDNS_ARCOUNT_SET(sldns_buffer_begin(c->buffer), 0);
+			sldns_buffer_flip(c->buffer);
```
```suggestion
			LDNS_ARCOUNT_SET(sldns_buffer_begin(c->buffer), 0);
			sldns_buffer_set_position(c->buffer, opt_rr_mark);
			sldns_buffer_flip(c->buffer);
```
Insert set_position, it is there in the other if statements, but not here, I figure you need it.

wcawijngaards commented on testcode/testpkts.c:
```diff
@@ -1428,13 +1505,32 @@ find_match(struct entry* entries, uint8_t* query_pkt, size_t len,
+			} else if(!p->match_ede_any &&
+				(uint16_t)info_code != p->ede_info_code) {
+				verbose(3, "bad EDE INFO-CODE. Expected: %d, "
+					"and got: %d\n", p->ede_info_code,
```
```suggestion
					"and got: %d\n", (int)p->ede_info_code,
```
Does this need a cast?

TCY16 replied:
can't hurt

wcawijngaards reviewed (commented):
This is a lot of stuff, but here are some remarks so you can see them. The review continues.

wcawijngaards commented on testdata/ede.tdir/bogus/make-broken-zone.sh:
```diff
@@ -0,0 +1,54 @@
+grep -v '	RRSIG	' tmp.signed > bogus/rrsig-failures.test.signed
+
+# cleanup
+rm -f base expired notyetincepted tmp.signed $CSK.*
```
Because of repeated use of the variable, the `$CSK.*` part removes only the last key and not the previous ones. This is harmless because the test runner script removes the test directory after the test is done. It makes a copy of the source test directory and files to run the test in. But the delete command does not seem do what the intent seems to be.

Commit:
process first round of review comments by @wcawijngaards

wcawijngaards reviewed (commented):
Comments at this point. Nice code :-)

Commit:
correctly remove broken keys in tdir script

wcawijngaards commented on testdata/ede.tdir/ede.dsc:
```diff
@@ -0,0 +1,16 @@
+BaseName: svcb
```
```suggestion
BaseName: ede
```
Fix the basename to match the tdir directory name.

wcawijngaards commented on testdata/ede.tdir/ede.post:
```diff
@@ -0,0 +1,31 @@
+# #-- fwd_tcp.pre--#
```
```suggestion
# #-- ede.pre--#
```
Fix comment.

wcawijngaards commented on testdata/ede.tdir/ede.post:
```diff
@@ -0,0 +1,31 @@
+[ -f ../.tpkg.var.master ] && source ../.tpkg.var.master
+# use .tpkg.var.test for in test variable passing
+[ -f .tpkg.var.test ] && source .tpkg.var.test
+
```
This file seems to contain fwd_tcp.pre contents, and not the ede.post contents. Where the test script stops the server that it started.

wcawijngaards commented on testdata/ede.tdir/ede.pre:
```diff
@@ -0,0 +1,32 @@
+# #-- fwd_tcp.pre--#
```
```suggestion
# #-- ede.pre--#
```
Fix comment.

wcawijngaards commented on testdata/ede.tdir/ede.pre:
```diff
@@ -0,0 +1,32 @@
+
+# start forwarder
+get_ldns_testns
+$LDNS_TESTNS -p $FWD_PORT fwd_tcp.testns >fwd.log 2>&1 &
```
The fwd_tcp.testns file is not committed. Also fwd_tcp.conf used a couple of lines later. Perhaps this test script simply does not work?

wcawijngaards commented on testdata/ede.tdir/ede.test:
```diff
@@ -0,0 +1,117 @@
+# #-- svcb.test --#
```
```suggestion
# #-- ede.test --#
```
Fix comment.

wcawijngaards commented on testdata/ede.tdir/ede.test:
```diff
@@ -0,0 +1,117 @@
+# use .tpkg.var.test for in test variable passing
+[ -f .tpkg.var.test ] && source .tpkg.var.test
+
+. ../common.sh
```
This file seems to do the start and stop of the server. But it does not stop the server if there is an error. The ede.pre and ede.post files contain garbage, and are also not listed in the .dsc file. Perhaps move the stop of the server to ede.post and list it so that the server is stopped when there is a failure in this test script. Or otherwise, in the exit 1 cases, also stop the started server. And remove or fix the ede.pre and ede.post files.

wcawijngaards commented on testdata/ede.tdir/ede.test:
```diff
@@ -0,0 +1,117 @@
+
+
+# start authoritative unbound in the background
+$PRE/unbound -d -c ub.conf > unbound.log 2>&1 &
```
If you start unbound here again, you need to stop the previous instance. Also you say it is the auhoritative server, but it uses ub.conf here, but ub2.conf before. And you store it in UNBOUND_PID2.

wcawijngaards commented on testdata/ede_localzone_dname_expansion.rpl:
```diff
@@ -0,0 +1,37 @@
+
+CONFIG_END
+
+SCENARIO_BEGIN Test ede-cache-shoop-not-authoritative
```
This comment is from another file, it should be about this test.

wcawijngaards reviewed (commented):
More comment cases.

wcawijngaards commented on util/config_file.c:
```diff
@@ -670,6 +672,7 @@ int config_set_option(struct config_file* cfg, const char* opt,
 	else if(strcmp(opt, "serve-expired-reply-ttl:") == 0)
 	{ IS_NUMBER_OR_ZERO; cfg->serve_expired_reply_ttl = atoi(val); SERVE_EXPIRED_REPLY_TTL=(time_t)cfg->serve_expired_reply_ttl;}
 	else S_NUMBER_OR_ZERO("serve-expired-client-timeout:", serve_expired_client_timeout)
+	else S_YNO("serve-expired-ede:", serve_expired_ede)
```
There is no do_ede added.

wcawijngaards commented on util/config_file.c:
```diff
@@ -1111,6 +1114,7 @@ config_get_option(struct config_file* cfg, const char* opt,
 	else O_YNO(opt, "serve-expired-ttl-reset", serve_expired_ttl_reset)
 	else O_DEC(opt, "serve-expired-reply-ttl", serve_expired_reply_ttl)
 	else O_DEC(opt, "serve-expired-client-timeout", serve_expired_client_timeout)
+	else O_YNO(opt, "serve-expired-ede", serve_expired_ede)
```
There is no do_ede added.

wcawijngaards commented on validator/val_utils.c:
```diff
@@ -775,7 +812,8 @@ val_dsset_isusable(struct ub_packed_rrset_key* ds_rrset)
 		else snprintf(aerr, sizeof(aerr), "%d",
-			(int)ds_get_key_algo(ds_rrset, 0));
+			(int)ds_get_key_algo(ds_rrset, i));
+
```
```suggestion
			(int)ds_get_key_algo(ds_rrset, 0));
```
Is this the result of a wrong merge? It seems to undo a commit that fixes a bug here by s/i/0/.

Commit:
make ede.tdir compatible with MacOS

Commit:
move from only using ede.tdir to using the .pre/.post structure

Commit:
fix correct scenario name

wcawijngaards reviewed (approved):
Code looks good! Hopefully the review fixes a couple of issues.

Commit:
process final round of comments by @wcawijngaards

Commit:
Merge branch 'master' into features/ede-basic

gthess commented on daemon/worker.c:
```diff
@@ -1030,32 +1054,167 @@ deny_refuse(struct comm_point* c, enum acl_access acl,
+			sldns_buffer_flip(c->buffer);
+			return 1;
+		} else
+			sldns_buffer_skip(c->buffer, 1); /* skip root label */
```
Nit: Add brackets for clarity since we already use them for the if part.
```suggestion
		} else {
			sldns_buffer_skip(c->buffer, 1); /* skip root label */
		}
```

gthess commented on doc/unbound.conf.5.in:
```diff
@@ -1770,6 +1774,13 @@ option can be used multiple times. The most specific match will be used.
+.TP 5
+.B do\-ede: \fI<yes or no>
+If enabled, Unbound will respond with Extended DNS Error codes (RFC8914).
+These EDEs attachinformative error messages to a response for various
```
Typo
```suggestion
These EDEs attach informative error messages to a response for various
```

gthess commented on services/localzone.c:
```diff
@@ -1338,6 +1339,13 @@ local_error_encode(struct query_info* qinfo, struct module_env* env,
 		rcode, edns, repinfo, temp, env->now_tv))
 		edns->opt_list_inplace_cb_out = NULL;
+
+	if (ede_code >= 0 /* >0 means a defined sldns_ede_code */
```
Typo?
```suggestion
	if(ede_code >= 0 /* >=0 means a defined sldns_ede_code */
```

gthess commented on services/localzone.c:
```diff
@@ -1533,9 +1541,11 @@ local_data_answer(struct local_zone* z, struct module_env* env,
 			if(newtargetlen > LDNS_MAX_DOMAINLEN) {
 				qinfo->local_alias = NULL;
-				local_error_encode(qinfo, env, edns, repinfo,
+				local_error_encode(qinfo, env, edns,repinfo,
```
Nit
```suggestion
				local_error_encode(qinfo, env, edns, repinfo,
```

gthess commented on services/mesh.c:
```diff
@@ -1295,6 +1295,22 @@ mesh_send_reply(struct mesh_state* m, int rcode, struct reply_info* rep,
+			rep->security == sec_status_secure_sentinel_fail)) {
+
+			char *reason = m->s.env->cfg->val_log_level >= 2
+			             ? errinf_to_str_bogus(&m->s) : NULL;
```
Formatting
```suggestion
			char *reason = m->s.env->cfg->val_log_level >= 2
				? errinf_to_str_bogus(&m->s) : NULL;
```

gthess commented on services/mesh.c:
```diff
@@ -1295,6 +1295,22 @@ mesh_send_reply(struct mesh_state* m, int rcode, struct reply_info* rep,
+			char *reason = m->s.env->cfg->val_log_level >= 2
+			             ? errinf_to_str_bogus(&m->s) : NULL;
+			sldns_ede_code reason_bogus = rep->reason_bogus != LDNS_EDE_DNSSEC_BOGUS
+			                            ? rep->reason_bogus : errinf_to_reason_bogus(&m->s);
```
Formatting
```suggestion
			sldns_ede_code reason_bogus = rep->reason_bogus != LDNS_EDE_DNSSEC_BOGUS
				? rep->reason_bogus : errinf_to_reason_bogus(&m->s);
```

gthess commented on services/mesh.c:
```diff
@@ -2050,6 +2068,13 @@ mesh_serve_expired_callback(void* arg)
 		}
 
+		/* Add EDE Stale Answer (RCF8914). Ignore global do_ede as this is
+		 * warning instead of an error */
```
I don't understand the warning/error logic. Also noted in worker.c; discussion should continue there but any changes should apply here as well.

gthess commented on daemon/worker.c:
```diff
@@ -716,15 +728,24 @@ answer_from_cache(struct worker* worker, struct query_info* qinfo,
-			qinfo, id, flags, edns);
+	} else {
+		/* We don't check the global do_ede as this is a warning, not
+		 * an error */
```
I don't understand the warning/error logic.
For me it seems that currently the options serve-expired-ede and do-ede are independent.
This means that there is no global on/off option for EDE. But it allows for serve expired EDE without the rest of the EDE cases.
Personally I would want do-ede to be global.
What do we think? (CC @wtoorop, @wcawijngaards )

TCY16 replied:
The reason we decided that `serve-expired-ede` would be independent is that it's technically not an error, but a warning. We weren't planning on adding any of the warning EDE's in the "basic" cases, but this specific one was asked for by a user during development. To distinguish between these semantically, we kept them as separate configurables.

gthess commented on services/mesh.c:
```diff
@@ -2050,6 +2068,13 @@ mesh_serve_expired_callback(void* arg)
+		 * warning instead of an error */
+		if (r->edns.edns_present && qstate->env->cfg->serve_expired_ede) {
+			edns_opt_list_append_ede(&r->edns.opt_list_out,
+				mstate->s.region, LDNS_EDE_STALE_ANSWER, "");
```
Why not NULL as in earlier cases in localzone.c?
```suggestion
			edns_opt_list_append_ede(&r->edns.opt_list_out,
				mstate->s.region, LDNS_EDE_STALE_ANSWER, NULL);
```

TCY16 replied:
both are fine, changing to NULL for consistency

gthess commented on util/config_file.h:
```diff
@@ -1282,6 +1291,13 @@ void errinf_dname(struct module_qstate* qstate, const char* str,
+ * Check the sldns_ede_code of the qstate.
+ * @param qstate: query state.
+ * @return LDNS_EDE_DNSSEC_BOGUS by default, or another sldns_ede_code 
+ * if this is set.
```
Update text
```suggestion
 * @return LDNS_EDE_DNSSEC_BOGUS by default, or the first explicitly set
 *   sldns_ede_code.
```

gthess commented on util/config_file.h:
```diff
@@ -799,6 +804,8 @@ struct config_strlist {
 	/** config option string */
 	char* str;
+	/** EDE code companion to the error str */
+	sldns_ede_code reason_bogus;
```
I would say to not overload a generic list of strings with the reason_bogus information.
`struct config_strlist` was used for convenience in `qstate->errinf` (because a list of strings was needed).
If we need an augmented list of strings with the added reason_bogus variable I would suggest to create one (e.g., errinf_strlist) in module.h so that modules and the qstate can use that one instead of the config_strlist one.

gthess commented on util/config_file.c:
```diff
@@ -2569,14 +2575,42 @@ void errinf(struct module_qstate* qstate, const char* str)
 	}
 	p->next = NULL;
 	p->str = regional_strdup(qstate->region, str);
+	p->reason_bogus = LDNS_EDE_DNSSEC_BOGUS;
```
I don't find this correct. errinf is called from other places (e.g., iterator.c) that have nothing to do with validation.

gthess commented on util/config_file.c:
```diff
@@ -2569,14 +2575,42 @@ void errinf(struct module_qstate* qstate, const char* str)
+}
+
+void errinf_ede(struct module_qstate* qstate,
+	const char* str, sldns_ede_code reason_bogus)
```
This is the same function as errinf above.
Could the extra reason_bogus be passed always with errinf and get rid of the duplication here?

gthess commented on util/data/msgreply.h:
```diff
@@ -528,7 +534,38 @@ void log_query_info(enum verbosity_value v, const char* str,
+		struct {						\
+			uint16_t code;					\
+			char text[sizeof(TXT) - 1];			\
+		} ede = { htons(CODE), TXT };				\
+                verbose(VERB_ALGO, "attached EDE code: %d with"		\
+                        " message: %s", CODE, TXT);			\
```
spaces -> tabs
```suggestion
		} ede = { htons(CODE), TXT };				\
		verbose(VERB_ALGO, "attached EDE code: %d with"		\
			" message: %s", CODE, TXT);			\
```

gthess commented on validator/val_sigcrypt.c:
```diff
@@ -1598,7 +1674,8 @@ dnskey_verify_rrset_sig(struct regional* region, sldns_buffer* buf,
 		 * rest is sure to be OK. */
-		if(!check_dates(ve, now, sig+2+8, sig+2+12, reason)) {
+		if(!check_dates(ve, now, sig+2+8, sig+2+12,
+					reason, reason_bogus)) {
```
Formatting
```suggestion
		if(!check_dates(ve, now, sig+2+8, sig+2+12,
			reason, reason_bogus)) {
```

gthess commented on validator/val_sigcrypt.c:
```diff
@@ -1410,6 +1461,8 @@ check_dates(struct val_env* ve, uint32_t unow,
 				incep, now);
 			*reason = "signature expired";
+			if (reason_bogus)
+				*reason_bogus = LDNS_EDE_SIGNATURE_EXPIRED;
```
Formatting
```suggestion
			if(reason_bogus)
				*reason_bogus = LDNS_EDE_SIGNATURE_EXPIRED;
```

gthess commented on validator/val_sigcrypt.c:
```diff
@@ -1397,6 +1446,8 @@ check_dates(struct val_env* ve, uint32_t unow,
 				" before inception date", expi, incep, now);
 			*reason = "signature before inception date";
+			if (reason_bogus)
+				*reason_bogus = LDNS_EDE_SIGNATURE_NOT_YET_VALID;
```
Formatting
```suggestion
			if(reason_bogus)
				*reason_bogus = LDNS_EDE_SIGNATURE_NOT_YET_VALID;
```

gthess commented on validator/val_sigcrypt.c:
```diff
@@ -1386,6 +1427,14 @@ check_dates(struct val_env* ve, uint32_t unow,
 		sigdate_error("verify: inception after expiration, "
 			"signature bad", expi, incep, now);
 		*reason = "signature inception after expiration";
+		if (reason_bogus){
```
Formatting
```suggestion
		if(reason_bogus){
```

gthess commented on validator/val_sigcrypt.c:
```diff
@@ -1361,8 +1402,8 @@ subtract_1982(uint32_t a, uint32_t b)
-check_dates(struct val_env* ve, uint32_t unow,
-	uint8_t* expi_p, uint8_t* incep_p, char** reason)
+check_dates(struct val_env* ve, uint32_t unow, uint8_t* expi_p,
+		uint8_t* incep_p, char** reason, sldns_ede_code *reason_bogus)
```
Formatting
```suggestion
check_dates(struct val_env* ve, uint32_t unow, uint8_t* expi_p,
	uint8_t* incep_p, char** reason, sldns_ede_code *reason_bogus)
```

gthess commented on validator/val_sigcrypt.c:
```diff
@@ -661,14 +699,17 @@ dnskeyset_verify_rrset_sig(struct module_env* env, struct val_env* ve,
 	}
 	if(numchecked == 0) {
 		*reason = "signatures from unknown keys";
+		if (reason_bogus)
```
Formatting
```suggestion
		if(reason_bogus)
```

gthess commented on validator/val_sigcrypt.c:
```diff
@@ -649,6 +685,8 @@ dnskeyset_verify_rrset_sig(struct module_env* env, struct val_env* ve,
 	int buf_canon = 0;
 	verbose(VERB_ALGO, "verify sig %d %d", (int)tag, algo);
 	if(!dnskey_algo_id_is_supported(algo)) {
+		if (reason_bogus)
```
Formatting
```suggestion
		if(reason_bogus)
```

gthess commented on validator/val_sigcrypt.c:
```diff
@@ -599,6 +622,16 @@ dnskey_verify_rrset(struct module_env* env, struct val_env* ve,
+}
+
+enum sec_status 
+dnskey_verify_rrset_ede(struct module_env* env, struct val_env* ve,
```
I would say that this function is not needed. We can just use `dnskey_verify_rrset` and pass NULL in the places where `*reason_bogus` is not relevant (there are only a few).

gthess commented on validator/val_sigcrypt.c:
```diff
@@ -551,12 +571,15 @@ dnskeyset_verify_rrset(struct module_env* env, struct val_env* ve,
 		if(algo_needs_num_missing(&needs) == 0) {
 			verbose(VERB_QUERY, "zone has no known algorithms");
 			*reason = "zone has no known algorithms";
+			if (reason_bogus)
```
Formatting
```suggestion
			if(reason_bogus)
```

gthess commented on validator/val_sigcrypt.c:
```diff
@@ -543,6 +561,8 @@ dnskeyset_verify_rrset(struct module_env* env, struct val_env* ve,
 		verbose(VERB_QUERY, "rrset failed to verify due to a lack of "
 			"signatures");
 		*reason = "no signatures";
+		if (reason_bogus)
```
Formatting
```suggestion
		if(reason_bogus)
```

gthess commented on validator/val_sigcrypt.c:
```diff
@@ -530,6 +530,24 @@ dnskeyset_verify_rrset(struct module_env* env, struct val_env* ve,
+dnskeyset_verify_rrset_ede(struct module_env* env, struct val_env* ve,
+	struct ub_packed_rrset_key* rrset, struct ub_packed_rrset_key* dnskey,
+	uint8_t* sigalg, char** reason, sldns_ede_code *reason_bogus,
+	sldns_pkt_section section, struct module_qstate* qstate)
```
I would say that this function is not needed. We can just use `dnskeyset_verify_rrset` and pass NULL in the places where `*reason_bogus` is not relevant (there are only a few).

TCY16 replied:
looking at the _ede and non-_ede versions, they seems to do different things as opposed to the other _ede functions you mentioned, which were just wrappers

gthess replied:
It seems like a wrapper to me (line 529), or am I confusing something?

gthess commented on validator/val_sigcrypt.c:
```diff
@@ -530,6 +530,24 @@ dnskeyset_verify_rrset(struct module_env* env, struct val_env* ve,
+	struct ub_packed_rrset_key* dnskey, size_t sig_idx, 
+	struct rbtree_type** sortree,
+	char** reason, sldns_ede_code *reason_bogus,
+	sldns_pkt_section section, struct module_qstate* qstate);
```
Now that this function is static, the declaration is cleaner to appear at the top of the file instead of here.

gthess commented on validator/val_utils.c:
```diff
@@ -571,15 +592,16 @@ val_verify_DNSKEY_with_DS(struct module_env* env, struct val_env* ve,
-	struct module_qstate* qstate)
-{
+	sldns_ede_code *reason_bogus, struct module_qstate* qstate)
+{	
```
Formatting
```suggestion
{
```

gthess commented on validator/val_utils.c:
```diff
@@ -617,6 +650,8 @@ val_verify_DNSKEY_with_TA(struct module_env* env, struct val_env* ve,
 		verbose(VERB_QUERY, "DNSKEY RRset did not match DS RRset "
 			"by name");
 		*reason = "DNSKEY RRset did not match DS RRset by name";
+		if (reason_bogus)
```
Formatting
```suggestion
		if(reason_bogus)
```

gthess commented on validator/val_utils.c:
```diff
@@ -625,6 +660,8 @@ val_verify_DNSKEY_with_TA(struct module_env* env, struct val_env* ve,
 		verbose(VERB_QUERY, "DNSKEY RRset did not match anchor RRset "
 			"by name");
 		*reason = "DNSKEY RRset did not match anchor RRset by name";
+		if (reason_bogus)
```
Formatting
```suggestion
		if(reason_bogus)
```

gthess commented on validator/val_utils.c:
```diff
@@ -393,6 +393,16 @@ enum sec_status
+val_verify_rrset_entry_ede(struct module_env* env, struct val_env* ve,
+        struct ub_packed_rrset_key* rrset, struct key_entry_key* kkey,
+	char** reason, sldns_ede_code *reason_bogus,
+	sldns_pkt_section section, struct module_qstate* qstate)
```
I would say that this function is not needed. We can just use `val_verify_rrset_entry` and pass NULL in the places where *reason_bogus is not relevant (there are only a few).

gthess commented on validator/val_utils.c:
```diff
@@ -488,11 +498,22 @@ int val_favorite_ds_algo(struct ub_packed_rrset_key* ds_rrset)
+val_verify_DNSKEY_with_DS_ede(struct module_env* env, struct val_env* ve,
+	struct ub_packed_rrset_key* dnskey_rrset,
+	struct ub_packed_rrset_key* ds_rrset, uint8_t* sigalg, char** reason,
+	sldns_ede_code *reason_bogus, struct module_qstate* qstate)
```
I would say that this function is not needed. We can just use `val_verify_DNSKEY_with_DS` and pass NULL in the places where *reason_bogus is not relevant (there are only a few).

gthess commented on validator/val_utils.c:
```diff
@@ -597,12 +619,23 @@ val_verify_new_DNSKEYs(struct regional* region, struct module_env* env,
+	struct ub_packed_rrset_key* dnskey_rrset,
+	struct ub_packed_rrset_key* ta_ds,
+	struct ub_packed_rrset_key* ta_dnskey, uint8_t* sigalg, char** reason,
+	sldns_ede_code *reason_bogus, struct module_qstate* qstate)
```
I would say that this function is not needed. We can just use `val_verify_DNSKEY_with_TA` and pass NULL in the places where *reason_bogus is not relevant (there are only a few).

gthess reviewed (changes requested):
Nice work! Here is the first part of the review (missing review of validator.c and all the test cases).
The review comments are mostly nits but there are also some suggestions and some changes that I would like to see.
I'll start working on the second part of the review.

Commit:
apply suggestions from @gthess's code review, as well as consolidating superfluous _ede functions in the validator

Commit:
add a few nits from @gthess's review, add LDNS_EDE_NONE as none-type and change logic accordingly, and change validator logic to be more careful with EDE overwrites and change logic accordingly

Commit:
Merge branch 'master' into features/ede-basic

gthess commented on testdata/black_key_entry.rpl:
```diff
@@ -577,7 +578,7 @@ ENTRY_END
 STEP 120 CHECK_ANSWER
 ENTRY_BEGIN
-MATCH all
+MATCH all ede=6
```
Maybe this answer needs to attach EDE 7 instead? errinf seem to also contain signature expired.

gthess commented on testdata/black_ds_entry.rpl:
```diff
@@ -595,7 +596,7 @@ ENTRY_END
 STEP 120 CHECK_ANSWER
 ENTRY_BEGIN
-MATCH all
+MATCH all ede=6
```
Maybe this answer needs to attach EDE 7 instead? errinf seem to also contain signature expired.

gthess commented on testdata/black_prime_entry.rpl:
```diff
@@ -304,7 +305,7 @@ ENTRY_END
 STEP 120 CHECK_ANSWER
 ENTRY_BEGIN
-MATCH all
+MATCH all ede=6
```
Maybe this answer needs to attach EDE 7 instead? errinf seem to also contain signature expired.

gthess commented on testdata/ede_cache_snoop_noth_auth.rpl:
```diff
@@ -0,0 +1,33 @@
+SCENARIO_BEGIN Test ede-cache-snoop-not-authoritative
+; Scenario overview:
+; - query for example.com. A record with EDNS
+; - check that we a refused answer with EDE (RFC8914) code 20 - Not Authoritative
```
```suggestion
; - query for example.com. A record with EDNS without the RD bit
; - check that we get a refused answer with EDE (RFC8914) code 20 - Not Authoritative
```

gthess commented on testdata/ede_acl_refused.rpl:
```diff
@@ -0,0 +1,35 @@
+SCENARIO_BEGIN Test ede-acl-refused
+; Scenario overview:
+; - query for example.com. A record with EDNS
+; - check that we a refused answer with EDE (RFC8914) code 18 - Prohibited
```
```suggestion
; - check that we get a refused answer with EDE (RFC8914) code 18 - Prohibited
```

gthess commented on testdata/ede_localzone_dname_expansion.rpl:
```diff
@@ -0,0 +1,37 @@
+SCENARIO_BEGIN Test ede-localzone-dname-expansion
+; Scenario overview:
+; - query for www.qhqwer.qwer.qwer.h.example.com. (a large Qname) A record with EDNS
+; - check that we a YXDOMAIN answer with EDE (RFC8914) code 0 - Other (with adds a DNAME expansion message)
```
```suggestion
; - check that we get a YXDOMAIN answer with EDE (RFC8914) code 0 - Other (which adds a DNAME expansion message)
```

gthess commented on testdata/local_transparent_sametype.rpl:
```diff
@@ -48,7 +48,7 @@ ENTRY_END
 STEP 10 CHECK_ANSWER
 ENTRY_BEGIN
-MATCH all
+MATCH all_noedns
```
I believe `all_noedns` is not needed here. Maybe remnant while developing?
```suggestion
MATCH all
```

gthess commented on testdata/local_typetransparent.rpl:
```diff
@@ -59,7 +59,7 @@ ENTRY_END
 STEP 10 CHECK_ANSWER
 ENTRY_BEGIN
-MATCH all
+MATCH all_noedns
```
I believe `all_noedns` is not needed here. Maybe remnant while developing?
```suggestion
MATCH all
```

gthess commented on testdata/serve_expired.rpl:
```diff
@@ -55,6 +56,8 @@ ENTRY_BEGIN
 	SECTION QUESTION
 		example.com. IN A
+		HEX_EDNSDATA_BEGIN
+	    HEX_EDNSDATA_END
```
Not needed here
```suggestion
```

gthess commented on testdata/serve_expired.rpl:
```diff
@@ -70,6 +73,8 @@ ENTRY_BEGIN
 	SECTION ADDITIONAL
 		ns.example.com. IN A 1.2.3.4
+		HEX_EDNSDATA_BEGIN
+	    HEX_EDNSDATA_END
```
Not needed here
```suggestion
```

gthess commented on testdata/serve_expired.rpl:
```diff
@@ -95,6 +102,8 @@ ENTRY_BEGIN
 	SECTION ADDITIONAL
 		ns.example.com. 30 IN A 1.2.3.4
+		HEX_EDNSDATA_BEGIN
+	    HEX_EDNSDATA_END
```
Not needed anymore
```suggestion
```

gthess commented on testdata/serve_expired.rpl:
```diff
@@ -103,6 +112,8 @@ ENTRY_BEGIN
 	SECTION QUESTION
 		example.com. IN A
+		HEX_EDNSDATA_BEGIN
+	    HEX_EDNSDATA_END
```
Not needed here
```suggestion
```

gthess commented on testdata/serve_expired.rpl:
```diff
@@ -117,6 +128,8 @@ ENTRY_BEGIN
 	SECTION ADDITIONAL
 		ns.example.com. IN A 1.2.3.4
+		HEX_EDNSDATA_BEGIN
+	    HEX_EDNSDATA_END
```
Not needed here
```suggestion
```

gthess commented on testdata/serve_expired_client_timeout.rpl:
```diff
@@ -60,6 +61,8 @@ ENTRY_BEGIN
 	SECTION QUESTION
 		example.com. IN A
+		HEX_EDNSDATA_BEGIN
+	    HEX_EDNSDATA_END
```
Not needed here
```suggestion
```

gthess commented on testdata/serve_expired_client_timeout.rpl:
```diff
@@ -75,6 +78,8 @@ ENTRY_BEGIN
 	SECTION ADDITIONAL
 		ns.example.com. IN A 1.2.3.4
+		HEX_EDNSDATA_BEGIN
+	    HEX_EDNSDATA_END
```
Not needed here
```suggestion
```

gthess commented on testdata/serve_expired_client_timeout.rpl:
```diff
@@ -122,6 +131,8 @@ ENTRY_BEGIN
 	SECTION ADDITIONAL
 		ns.example.com. 3600 IN A 1.2.3.4
+		HEX_EDNSDATA_BEGIN
+	    HEX_EDNSDATA_END
```
Not needed here
```suggestion
```

gthess commented on testdata/serve_expired.rpl:
```diff
@@ -80,12 +85,14 @@ STEP 30 QUERY
 	SECTION QUESTION
 		example.com. IN A
+		HEX_EDNSDATA_BEGIN
+	    HEX_EDNSDATA_END
```
Having the DO bit is enough for EDNS
```suggestion
	REPLY DO
	SECTION QUESTION
		example.com. IN A
```

gthess commented on testdata/serve_expired.rpl:
```diff
@@ -80,12 +85,14 @@ STEP 30 QUERY
 ENTRY_BEGIN
-	MATCH all ttl
+	MATCH all ttl ede=3
 	REPLY QR RA NOERROR
```
We need the DO bit here now
```suggestion
	MATCH all ttl ede=3
	REPLY QR RA DO NOERROR
```

gthess commented on testdata/serve_expired_client_timeout.rpl:
```diff
@@ -86,6 +91,8 @@ ENTRY_BEGIN
 	SECTION QUESTION
 		example.com. IN A
+		HEX_EDNSDATA_BEGIN
+	    HEX_EDNSDATA_END
```
We can just put the DO bit
```suggestion
	REPLY RD DO
	SECTION QUESTION
		example.com. IN A
```

gthess commented on testdata/serve_expired_client_timeout.rpl:
```diff
@@ -104,6 +111,8 @@ ENTRY_BEGIN
 	SECTION ADDITIONAL
 		ns.example.com. 123 IN A 1.2.3.4
+		HEX_EDNSDATA_BEGIN
+	    HEX_EDNSDATA_END
```
Not needed anymore
```suggestion
```

gthess commented on testdata/serve_expired_client_timeout.rpl:
```diff
@@ -94,7 +101,7 @@ STEP 31 TIME_PASSES ELAPSE 1
 ENTRY_BEGIN
-	MATCH all ttl
+	MATCH all ttl ede=3
 	REPLY QR RD RA NOERROR
```
We need the DO bit here now
```suggestion
	MATCH all ttl ede=3
	REPLY QR RD RA DO NOERROR
```

gthess commented on testdata/serve_expired_reply_ttl.rpl:
```diff
@@ -54,12 +55,15 @@ ENTRY_BEGIN
 		example.com. IN A
+	SECTION ADDITIONAL
+    HEX_EDNSDATA_BEGIN
+    HEX_EDNSDATA_END
```
Not needed
```suggestion
```

gthess commented on testdata/serve_expired_reply_ttl.rpl:
```diff
@@ -54,12 +55,15 @@ ENTRY_BEGIN
 STEP 10 CHECK_ANSWER
 ENTRY_BEGIN
-	MATCH all ttl
+	MATCH all_noedns ttl
```
We can just do `all` now that we don't care about edns
```suggestion
	MATCH all ttl
```

gthess commented on testdata/serve_expired_reply_ttl.rpl:
```diff
@@ -69,6 +73,8 @@ ENTRY_BEGIN
 	SECTION ADDITIONAL
 		ns.example.com. IN A 1.2.3.4
+	HEX_EDNSDATA_BEGIN
+    HEX_EDNSDATA_END
```
Not needed
```suggestion
```

gthess commented on testdata/serve_expired_reply_ttl.rpl:
```diff
@@ -95,6 +101,8 @@ ENTRY_BEGIN
 	SECTION ADDITIONAL
 		ns.example.com. 123 A 1.2.3.4
+		HEX_EDNSDATA_BEGIN
+	    HEX_EDNSDATA_END
```
Not needed
```suggestion
```

gthess commented on testdata/serve_expired_servfail.rpl:
```diff
@@ -71,6 +72,8 @@ ENTRY_BEGIN
 	SECTION QUESTION
 		example.com. IN A
+		HEX_EDNSDATA_BEGIN
+	    HEX_EDNSDATA_END
```
Not needed
```suggestion
```

gthess commented on testdata/serve_expired_servfail.rpl:
```diff
@@ -86,6 +89,8 @@ ENTRY_BEGIN
 	SECTION ADDITIONAL
 		ns.example.com. IN A 1.2.3.4
+		HEX_EDNSDATA_BEGIN
+	    HEX_EDNSDATA_END
```
Not needed
```suggestion
```

gthess commented on testdata/serve_expired_servfail.rpl:
```diff
@@ -97,12 +102,14 @@ ENTRY_BEGIN
 	SECTION QUESTION
 		example.com. IN A
+		HEX_EDNSDATA_BEGIN
+	    HEX_EDNSDATA_END
```
Just the DO bit suffices
```suggestion
	REPLY RD DO
	SECTION QUESTION
		example.com. IN A
```

gthess commented on testdata/serve_expired_servfail.rpl:
```diff
@@ -97,12 +102,14 @@ ENTRY_BEGIN
 ENTRY_BEGIN
-	MATCH all ttl
+	MATCH all ttl ede=3
 	REPLY QR RD RA NOERROR
```
Do bit
```suggestion
	REPLY QR RD RA DO NOERROR
```

gthess commented on testdata/serve_expired_servfail.rpl:
```diff
@@ -112,6 +119,8 @@ ENTRY_BEGIN
 	SECTION ADDITIONAL
 		ns.example.com. 123 IN A 1.2.3.4
+		HEX_EDNSDATA_BEGIN
+	    HEX_EDNSDATA_END
```
Not needed
```suggestion
```

gthess commented on testdata/serve_expired_zerottl.rpl:
```diff
@@ -79,6 +80,8 @@ ENTRY_BEGIN
 	SECTION QUESTION
 		example.com. IN A
+		HEX_EDNSDATA_BEGIN
+	    HEX_EDNSDATA_END
```
Not needed
```suggestion
```

gthess commented on testdata/serve_expired_zerottl.rpl:
```diff
@@ -94,6 +97,8 @@ ENTRY_BEGIN
 	SECTION ADDITIONAL
 		ns.example.com. 0 IN A 1.2.3.4
+		HEX_EDNSDATA_BEGIN
+	    HEX_EDNSDATA_END
```
Not needed
```suggestion
```

gthess commented on testdata/serve_expired_zerottl.rpl:
```diff
@@ -105,6 +110,8 @@ ENTRY_BEGIN
 	SECTION QUESTION
 		example.com. IN A
+		HEX_EDNSDATA_BEGIN
+	    HEX_EDNSDATA_END
```
Not needed
```suggestion
```

gthess commented on testdata/serve_expired_zerottl.rpl:
```diff
@@ -120,6 +127,8 @@ ENTRY_BEGIN
 	SECTION ADDITIONAL
 		ns.example.com. 10 IN A 1.2.3.4
+		HEX_EDNSDATA_BEGIN
+	    HEX_EDNSDATA_END
```
Not needed
```suggestion
```

gthess commented on testdata/serve_expired_zerottl.rpl:
```diff
@@ -131,12 +140,14 @@ ENTRY_BEGIN
 	SECTION QUESTION
 		example.com. IN A
+		HEX_EDNSDATA_BEGIN
+	    HEX_EDNSDATA_END
```
DO bit
```suggestion
	REPLY RD DO
	SECTION QUESTION
		example.com. IN A
```

gthess commented on testdata/serve_expired_zerottl.rpl:
```diff
@@ -146,6 +157,8 @@ ENTRY_BEGIN
 	SECTION ADDITIONAL
 		ns.example.com. 123 IN A 1.2.3.4
+		HEX_EDNSDATA_BEGIN
+	    HEX_EDNSDATA_END
```
Not needed anymore
```suggestion
```

gthess commented on testdata/serve_expired_zerottl.rpl:
```diff
@@ -131,12 +140,14 @@ ENTRY_BEGIN
 ENTRY_BEGIN
-	MATCH all ttl
+	MATCH all ttl ede=3
 	REPLY QR RD RA NOERROR
```
DO bit
```suggestion
	MATCH all ttl ede=3
	REPLY QR RD RA DO NOERROR
```

gthess commented on testdata/serve_original_ttl.rpl:
```diff
@@ -113,12 +114,14 @@ ENTRY_BEGIN
 	SECTION QUESTION
 		example.com. IN A
+		HEX_EDNSDATA_BEGIN
+	    HEX_EDNSDATA_END
```
DO bit
```suggestion
	REPLY DO
	SECTION QUESTION
		example.com. IN A
```

gthess commented on testdata/serve_original_ttl.rpl:
```diff
@@ -113,12 +114,14 @@ ENTRY_BEGIN
 ENTRY_BEGIN
-	MATCH all ttl
+	MATCH all ttl ede=3
 	REPLY QR RA NOERROR
```
DO bit
```suggestion
	REPLY QR RA DO NOERROR
```

gthess commented on testdata/serve_original_ttl.rpl:
```diff
@@ -128,6 +131,8 @@ ENTRY_BEGIN
 	SECTION ADDITIONAL
 		ns.example.com. A 1.2.3.4
+		HEX_EDNSDATA_BEGIN
+	    HEX_EDNSDATA_END
```
Not needed anymore
```suggestion
```

gthess commented on testdata/val_faildnskey.rpl:
```diff
@@ -160,11 +161,12 @@ ENTRY_END
 SECTION QUESTION
 www.example.com. IN A
 SECTION ANSWER
+SECTION ADDITIONAL
```
Not needed
```suggestion
```

gthess commented on testdata/val_nsec3_nods_badsig.rpl:
```diff
@@ -226,13 +227,11 @@ ENTRY_END
 STEP 10 CHECK_ANSWER
 ENTRY_BEGIN
-MATCH all
+MATCH all ede=9
```
I believe this answer should return "signature expired".
NSEC3 functions (like `list_is_secure` in this case) do not seem to pass `reason_bogus` to validity functions and eventually `val_sigrypt.c::check_dates`

gthess commented on testdata/ede.tdir/ede.test:
```diff
@@ -0,0 +1,79 @@
+	echo "query to servfail.nl should return EDE signature expired or DNSKEY missing"
+	cat servfail.txt
+	exit 1
+fi
```
This test relies on an external domain which may cause problems in the future.
You could either create such zones (Signature Expired / DNSKEY Missing) for the test here or remove altogether since those are already tested in the `.rpl` files.

gthess committed:
EDE review: properly use LDNS_EDE_NONE in localzone.c.

gthess committed:
EDE review: document reason_bougs for dnskeyset_verify_rrset().

gthess commented on testdata/ede.tdir/ede.conf:
```diff
@@ -0,0 +1,54 @@
+
+	# @TODO change this to local file
+	#auto-trust-anchor-file: "/var/lib/unbound/root.key"
+	trust-anchor-file: "bogus/trust-anchors"
```
It seems the TODO is now done
```suggestion
	trust-anchor-file: "bogus/trust-anchors"
```

gthess reviewed (changes requested):
Finished the second part of the review.
Mostly nits (with clickable suggestions) but I believe a couple of tests should expect a different EDE value.

TCY16 committed:
Apply click suggestions from @gthess's code review

Co-authored-by: gthess <george@nlnetlabs.nl>

Commit:
change LDNS_EDE_NONE to be outside of the EDE range, correct nsec3 ede logic, remove superfluous test cases

Commit:
store ede code in key_entry_data struct to remember why key is bad, change coresponding tests

Commit:
remove superfluous check in val_kentry:key_entry_get_reason_bogus()

Commit:
fix typo in doc

gthess reviewed (changes requested):
Seems good!
Only thing that is pending is changing `serve-expired-ede` to `ede-serve-expired` and checking the global ede option together with that option.
Maybe also renaming the global option to just `ede` for easier grouping with the (future) ede options/toggles as discussed.

Commit:
rename cfg serve-expired-ede to ede-serve-expired and add add checking the global ede option together with that option

Commit:
document the ede-serve-expired changes in the manpage

Commit:
rename do-ede option to do ede

gthess committed:
EDE review: update man page.

gthess committed:
EDE review: update example.conf.in

gthess reviewed (approved):
LGTM, ready for merging!
